sktime/split/expandingcutoff.py

```diff
@@ -0,0 +1,157 @@
+#!/usr/bin/env python3 -u
+# copyright: sktime developers, BSD-3-Clause License (see LICENSE file)
+"""Splitter that successively expands the training window."""
+
+__author__ = ["ninedigits"]
+
+__all__ = [
+    "ExpandingCutoffSplitter",
+]
+
+from typing import Optional
+
+import numpy as np
+import pandas as pd
+
+from sktime.split.base import BaseSplitter
+from sktime.split.base._common import ACCEPTED_Y_TYPES, _check_fh, _inputs_are_supported
+from sktime.utils.validation.forecasting import check_step_length
+
+
+def _is_all_args_periodlike(args):
+    for arg in args:
+        if not isinstance(arg, pd.Period):
+            return False
+    return True
+
+
+class ExpandingCutoffSplitter(BaseSplitter):
+    """
+    Expanding cutoff splitter for time series data.
+
+    This splitter combines elements of `ExpandingWindowSplitter` and `CutoffSplitter`
+    to create training and testing sets. Unlike `ExpandingWindowSplitter` which begins
+    with a fixed initial window, this splitter uses a specific cutoff point as the
+    starting window for the training set. The training set then expands incrementally
+    in each split until it reaches the end of the series.
+
+    The test set is defined by a forecast horizon relative to the last point in the
+    training set, containing as many subsequent indices as specified by the `fh`
+    parameter.
+                          c
+                          ↓
+    |---------------------|←---fh---→|------|
```

Comment:
@fkiraly Any issues with including characters like `↓←→` as seen in line 42-43? Not sure if this is causing one of the unit tests to fail.

Reply:
I am not sure - I cannot download or access the logs. So, perhaps? I will try to run the tests locally.

@yarnabrina or @benHeid - can you see the test logs?

Reply:
> @yarnabrina or @benHeid - can you see the test logs?

Yes, I can see CI logs. Sharing two sample failures:

> FAILED sktime/tests/test_all_estimators.py::TestAllObjects::test_set_params_sklearn[ExpandingCutoffSplitter] - AssertionError: get_params result of ExpandingCutoffSplitter (x) does not match what was passed to set_params (y). Reason for discrepancy: [fh].type, x.type = <class 'sktime.forecasting.base._fh.ForecastingHorizon'> != y.type = <class 'list'>
assert False
> FAILED sktime/split/tests/test_expandingcutoff.py::test_expandingcutoff_fh_list_007 - AttributeError: attribute 'freq' of 'pandas._libs.tslibs.timestamps._Timestamp' objects is not writable

Reply:
Thanks @yarnabrina and @fkiraly. I believe I fixed those test cases, I think a few more errors might pop up for other issues. 

I think L119-L127
```python3
        input_args = [y[0], self.cutoff]
        if not _inputs_are_supported(input_args) and not _is_all_args_periodlike(
            input_args
        ):
            msg = (
                "y indicies and cutoff must have the same datatypes, but instead "
                f"found type(y) = {y.dtype} and type(cutoff) = {type(self.cutoff)}"
            )
            raise TypeError(msg)
```
Might conflict with:

```python3
    def get_test_params(cls, parameter_set="default"):
        """..."""
        return [{"fh": [2, 4], "cutoff": 3, "step_length": 1}]
```

I have the current class designed such that you can only use `cutoff` as an int if the index of `y` is also int, and conversely, when `y` has a datelike index, then `cutoff` must be datelike.

Reply:
Hm, I see - this links to a general issue, where the current test system is unable to select - or skip - test scenarios based on data.

What we can do, however, is work with tags. If we introduce a new tag for the kind of index the estimator can work with, we can add some selection logic.

Not sure though if this would be overkill - an alternative option would be having only test parameter settings that work with all types of indices, and test others separately.

Can you explain in detail why integer `fh` does not work with all data types?

Reply:
> If you have understood this through debugging: can you explain to me again, in detail, why integer fh does not work with all index types?

Actually, the types of the `y` index and `cutoff` must match. We use the `cutoff` parameter to determine the point within `y` at which to divide the data into training and testing sets. Comparing a datetime value with an integer would result in an error. Implementing a type check could help users identify this issue beforehand. But without the type check, I think the `get_test_params` would still eventually throw and error because we'd need a way of telling it only to use datetime-like y indicies with datetime-like cutoffs.

Reply:
this reminds me of this discussion with @hazrulakmal about index vs fh/cutoff combinations:
https://github.com/sktime/sktime/pull/5009#issuecomment-1696089996

I think we should be consistent with the cases in there, although I do not fully remember the final API design outcome.

E.g., if `cutoff` is int and `y.index` is not, then we would interpret `cutoff` as an `iloc` reference.

Reply:
sorry that this is getting more complicated than thought. The interaction of indices in the params and indices of the data seems a bit unpleasant...

Reply:
This method could be effective, but it might lead to unexpected behaviors from the splitter when dealing with hierarchical data. Consider the following: if we have multiple time series and employ the `ExpandingCutoffSplitter` with a common index, users must understand how the indices are aligned. For instance, do all three time series start at index 1, or do they conclude at the same index, as demonstrated in this example? They don’t necessarily need to end on the same index, but users would need to have designed the input data such that the relationships between indices are established. The primary purpose of the `ExpandingCutoffSplitter` is to use a specific index point to begin testing, which is particularly useful for forecasting multiple time series using a common cutoff date. Does this make sense? I might be overanalyzing it. 

```
index: 123456789
s1:    --*******
s2:    *********
s3:    -********
```

Reply:
Here's one of the other errors that was showing up. I think it might be do to the use of special characters in the docstring:

> ERROR sktime/split/tests/test_all_splitters.py::TestAllSplitters - sktime.utils._testing._conditional_fixtures.FixtureGenerationError: fixture estimator_instance failed to generate. 'charmap' codec can't decode byte 0x90 in position 1657: character maps to <undefined>
= 12853 passed, 143 skipped, 1 xfailed, 569 warnings, 4 errors in 1371.93s (0:22:51) =

Reply:
yes, that's probably related to one of the special characters - it is always a bit of a hassle to decode "position 1234" to the actual location.

Reply:
> Does this make sense? I might be overanalyzing it.

I think it makes sense, and it might be confusing to the user if not explained clearly how cutoff and index types combine.

Still, I can think of `iloc` indexing in case of date-like index making sense: In the picture above:

index: 123456789
s1:    --*******
s2:    *********
s3:    -********
```

this alignment happens if you use negative `iloc` indices, e.g., -1 and -2 will index the last two indices.

Reply:
> this alignment happens if you use negative iloc indices, e.g., -1 and -2 will index the last two indices.

I’d considered this as well. I'd be open to modifying the code to work with datetime indices when the cutoff value is an int. Should we design it such that only negative cutoff values are allowed when working with datetime-like indices, or should we allow all integers? The former might still be a headache for the `get_test_params` method.

We need to clarify some assumptions. For hierarchical data, it’s not guaranteed that subgroups will end with the same index. If they don't end with the same index (represented by a datetime-like value in this example), then negative indices would reference different datetime values. However, if they do end with the same index, which I suspect is the case for most scenarios, then negative indices should work appropriately.

What do you think?

Reply:
> Should we design it such that only negative cutoff values are allowed when working with datetime-like indices, or should we allow all integers?

This is really a difficult question, sorry for the delay in replying.

I would allow all integers, even in date-like case. That would be consistent with negative integers, so needs less case handling.

> If they don't end with the same index (represented by a datetime-like value in this example), then negative indices would reference different datetime values. However, if they do end with the same index, which I suspect is the case for most scenarios, then negative indices should work appropriately.

I think it is fine if negative `iloc` index refers to different `loc` index - this could be a reasonable scenario, similar to the greedy splitter. In any case, as long as we explain this clearly to the user, it will be ok imo, all solutions seem to have an unpleasant aspect, but this is possibly the least confusing one.

Reply:
@fkiraly I actually implemented this in the last commit. The logic goes like this:

- When an index with dates is paired with an integer cutoff, the cutoff serves as an index locator (iloc).
- When the index is an integer, it directly compares the values, except when the integer is negative, in which case it functions as an iloc.
- example: index = 5 6 7 8; cutoff = 6 would be a direct comparison, whereas cutoff -2 would be the iloc equivalent. If, for example, someone used a cutoff of 10, then it would raise an error, since it's not present in the index.